File: Affine/Rust/affine.rs
#[derive(Default)]
pub(crate) struct Affine {
    a: usize,
    b: usize,
    pub input: String
}

impl Affine {
    pub fn new(a: usize, b: usize, input: String) -> Self {
        if Self::gcd(a, 26) != 1 {
            panic!("Error! 'a' must be coprime with 26");
        }

        Self { a, b, input }
    }

    fn gcd(a: usize, b: usize) -> usize {
        if b == 0 {
            return a;
        }
        Self::gcd(b, a % b)
    }

    pub(crate) fn encode(&self) -> String {
        let mut output = String::new();

        for c in self.input.chars() {
            if c.is_ascii_lowercase() {
                let encoded = ((c as usize - 'a' as usize) * self.a + self.b) % 26 + 'a' as usize;
                output.push(encoded as u8 as char);
            } else if c.is_ascii_uppercase() {
                let encoded = ((c as usize - 'A' as usize) * self.a + self.b) % 26 + 'A' as usize;
                output.push(encoded as u8 as char);
            } else if c.is_whitespace() || c.is_ascii_punctuation() {
                output.push(c);
            } else {
                panic!("Error! Unknown symbol: {}", c);
            }
        }

        output
    }

    fn mod_inverse(mut a: i32, m: i32) -> i32 {
        a = a % m;

        for i in 1..m {
            if a * i % m == 1 {
                return i;
            }
        }
        panic!("Error! No modular inverse")
    }

    pub(crate) fn decode(&self, encoded_version: &String) -> String {
        let mut output = String::new();
        let a_inv = Self::mod_inverse(self.a as i32, 26);

        for c in encoded_version.chars() {
            if c.is_ascii_lowercase() {
                let decoded = (a_inv * (c as i32 - 'a' as i32 - self.b as i32 + 26) % 26) + 'a' as i32;
                output.push(decoded as u8 as char);
            } else if c.is_ascii_uppercase() {
                let decoded = (a_inv * (c as i32 - 'A' as i32 - self.b as i32 + 26) % 26) + 'A' as i32;
                output.push(decoded as u8 as char);
            } else if c.is_whitespace() || c.is_ascii_punctuation() {
                output.push(c);
            } else {
                panic!("Error! Unknown symbol: {}", c);
            }
        }

        output
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encode_decode_basic() {
        let affine = Affine::new(5, 8, "hello".to_string());
        let encoded = affine.encode();
        let decoded = affine.decode(&encoded);
        assert_eq!(decoded, "hello");
    }

    #[test]
    fn test_preserves_case() {
        let affine = Affine::new(5, 8, "HelloWorld".to_string());
        let encoded = affine.encode();
        let decoded = affine.decode(&encoded);
        assert_eq!(decoded, "HelloWorld");
    }

    #[test]
    fn test_ignores_spaces_and_punctuation() {
        let affine = Affine::new(7, 3, "attack at dawn!".to_string());
        let encoded = affine.encode();
        let decoded = affine.decode(&encoded);
        assert_eq!(decoded, "attack at dawn!");
    }

    #[test]
    #[should_panic(expected = "Error! 'a' must be coprime with 26")]
    fn test_invalid_a_panics() {
        let _ = Affine::new(13, 5, "test".to_string());
    }

    #[test]
    #[should_panic(expected = "Error! Unknown symbol")]
    fn test_invalid_symbol_panics() {
        let affine = Affine::new(5, 8, "héllo".to_string());
        let _ = affine.encode();
    }

    #[test]
    fn test_different_keys() {
        let affine1 = Affine::new(7, 2, "rust".to_string());
        let encoded1 = affine1.encode();
        let decoded1 = affine1.decode(&encoded1);
        assert_eq!(decoded1, "rust");

        let affine2 = Affine::new(11, 15, "AffineCipher".to_string());
        let encoded2 = affine2.encode();
        let decoded2 = affine2.decode(&encoded2);
        assert_eq!(decoded2, "AffineCipher");
    }
}

File: Affine/C/Affine.c
#include "affine.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

static int gcd(const int a, const int b)
{
    return b == 0 ? a : gcd(b, a % b);
}

static int mod_inverse(int a)
{
    a = a % 26;
    for (int x = 1; x < 26; x++) {
        if (a * x % 26 == 1) {
            return x;
        }
    }
    fprintf(stderr, "No modular inverse exists for given 'a' = %d\n", a);
    exit(EXIT_FAILURE);
}

Affine affine_create(char *input, const int a, const int b)
{
    if (gcd(a, 26) != 1)
    {
        fprintf(stderr, "Parameter 'a' must be coprime with 26\n");
        exit(EXIT_FAILURE);
    }

    Affine cipher;
    cipher.a = a;
    cipher.b = b;
    cipher.input = input;

    for (char c = 'a'; c <= 'z'; c++)
    {
        printf("%c ", c);
    }
    printf("\n");

    return cipher;
}

char* affine_encode(const Affine *affine)
{
    const size_t len = strlen(affine->input);
    char *output = malloc(len + 1);
    if (!output)
    {
        perror("malloc");
        exit(EXIT_FAILURE);
    }

    for (size_t i = 0; i < len; i++)
    {
        const char c = affine->input[i];
        if (c >= 'a' && c <= 'z')
        {
            output[i] = (char)(((c - 'a') * affine->a + affine->b) % 26 + 'a');
        }
        else if (c >= 'A' && c <= 'Z')
        {
            output[i] = (char)(((c - 'A') * affine->a + affine->b) % 26 + 'A');
        }
        else if (isspace(c) || isdigit(c) || ispunct(c))
        {
            output[i] = c;
        }
        else
        {
            fprintf(stderr, "Invalid character: %c\n", c);
            free(output);
            exit(EXIT_FAILURE);
        }
    }
    output[len] = '\0';
    return output;
}

char* affine_decode(const Affine *affine, const char *encoded)
{
    const size_t len = strlen(encoded);
    char *output = malloc(len + 1);
    
    if (!output)
    {
        perror("malloc");
        exit(EXIT_FAILURE);
    }

    const int a_inv = mod_inverse(affine->a);

    for (size_t i = 0; i < len; i++)
    {
        const char c = encoded[i];
        if (c >= 'a' && c <= 'z')
        {
            output[i] = (char)((a_inv * (c - 'a' - affine->b + 26)) % 26 + 'a');
        }
        else if (c >= 'A' && c <= 'Z')
        {
            output[i] = (char)((a_inv * (c - 'A' - affine->b + 26)) % 26 + 'A');
        }
        else if (isspace(c) || isdigit(c) || ispunct(c))
        {
            output[i] = c;
        }
        else
        {
            fprintf(stderr, "Invalid character: %c\n", c);
            free(output);
            exit(EXIT_FAILURE);
        }
    }
    output[len] = '\0';
    return output;
}
